### ISSUE TITLE
ApolloClient onError doesn't catch mutation errors anymore

### ISSUE THREAD

**Issue by strblr** [apollographql/apollo-client#5708]
Here is my client configuration using Apollo Boost : 

```javascript
new ApolloClient({
  uri: `${window.location.origin}/api`,
  cache: new InMemoryCache(),
  onError({ graphQLErrors, networkError }) {
    if(graphQLErrors)
      graphQLErrors.forEach(error => notification.error({
        message: 'Error',
        description: error.message
      }))
    if(networkError)
      notification.error({
        message: 'Network Error',
        description: `A network error has occurred. Please check out your connection.`
      })
  },
  request(operation) {
    const currentUser = readStore('currentUser')
    currentUser && operation.setContext({
      headers: { authorization: currentUser.token }
    })
  }
})
```

**Intended outcome:**
Before the bug, the onError callback used to catch errors just fine.

**Actual outcome:**
Errors are said to remain uncaught and cause apps (using CRA in my case) to crash.

**How to reproduce the issue:**
Try switching to the latest version of apollo-boost and just throw a Graphql error on the server.

**Versions**
React : 16.12.0
Apollo-boost : 0.4.7

**Comment by Gallevy** [comment]
for me query errors are fine and mutation errors throw uncaught exception, can you confirm if you have the same problem?

**Comment by ostrebler** [comment]
> for me query errors are fine and mutation errors throw uncaught exception, can you confirm if you have the same problem?

Yep, same for me.

**Comment by viiiprock** [comment]
I think I got the same thing as `Uncaught (in promise) Error`

**Comment by viiiprock** [comment]
I think I got the same thing as `Uncaught (in promise) Error` logged.

**Comment by viiiprock** [comment]
But...,finally, I resolved by adding `catch` in promise mutation, for example. You have:
`const [login] = useMutation(LOGIN)`

then when you call mutation by an event

```js
login({
 variables: {
   input: { email, password }
 }
}).catch(err => console.error(err))
```

Our issue may resolve but `onError` issue is still there so it still need to be fixed.

**Comment by Alex-Yakubovsky** [comment]
I'm having a similar issue. When the response from a mutation contains an error I get an error that there was an unhandled rejection. This slows development since `react-dev-server` displays an error overlay that you have to close. Errors from `useQuery` and `useLazyQuery` are handled differently, they get caught internally by `apollo-client`.  At the moment I'm doing 

```js
const [cb, {data, error}] = useMutation(MUTATION, {onError: () => null})
```

but that's not ideal either. I was previously using apollo-boost and this was not an issue.

**Comment by pfarnach** [comment]
I'm also seeing this issue. Any mutation-generated network error throws unhandled exceptions, while query-generated errors do not.  My `onError` method is similar to the one above, but I'm not using Apollo Boost.  Adding a `catch` block to the mutation "fixes" the issue but is a pretty gross solution.

Any word on this from the maintainers? This seems like a pretty big regression.

```
"apollo-client": "2.6.8"
"apollo-link-error": "1.1.12"
"react": "16.12.0"
```

**Comment by gentleShark** [comment]
also seeing the same problem. @viiiprock thanks for the workaround.

**Comment by kissthom** [comment]
Same issue here.

If I had to guess I'd say this line in the catch block makes the trouble.   
[MutationData.runMutation](https://github.com/apollographql/apollo-client/blob/0540f98af9e46c5cccbd5b5cd4bd03bc6b915aab/src/react/data/MutationData.ts#L76)

````ts
  private runMutation = (
    mutationFunctionOptions: MutationFunctionOptions<
      TData,
      TVariables
    > = {} as MutationFunctionOptions<TData, TVariables>
  ) => {
    this.onMutationStart();
    const mutationId = this.generateNewMutationId();

    return this.mutate(mutationFunctionOptions)
      .then((response: FetchResult<TData>) => {
        this.onMutationCompleted(response, mutationId);
        return response;
      })
      .catch((error: ApolloError) => {
        this.onMutationError(error, mutationId);
        if (!this.getOptions().onError) throw error; // !!! I mean this line
      });
  };
// ... more code here...
  private onMutationError(error: ApolloError, mutationId: number) {
    const { onError } = this.getOptions();

    if (this.isMostRecentMutation(mutationId)) {
      this.updateResult({
        loading: false,
        error,
        data: undefined,
        called: true
      });
    }

    if (onError) {
      onError(error); 
    }
  }
````

Of course I don't know the whole implementation but it seems to me that in MutationData.onMutationError the errors are already handled pretty well, onError will be called conditionally as it should be so I'm not sure if rethrowing the error is really necessary.

**Comment by schester44** [comment]
Any update on this? I have a pretty large app that relied on this pattern to minimize the amount of try/catch and .catch that was needed. Made for a super clean code base knowing I could rely on apollo-link-error when needed.


Not sure what changed, guess I need to look at my past few commits, but lately all mutation errors throw an unhandled error.

Is this change documented in a change log somewhere?

**Comment by LA** [comment]
Would like to add as well, that seemingly out of nowhere, my useQuery is no longer calling the global onError I defined using apollo-boost client. I can only access the error on the error property of useQuery or passing onError directly to the useQuery options. 

However, since onError on useQuery only provides an ApolloError as opposed to an ErrorResponse, this completely dismantles my token refreshing system which relied on the ability to forward requests after refreshing token along the lines of https://www.apollographql.com/docs/link/links/error/#retrying-failed-requests

Am I wrong in my understanding that this is how apollo-link onError is supposed to be used?

**Comment by bravecow** [comment]
apollo v3.1.3 confirm the same error

**Comment by bkiac** [comment]
Why is there no activity on this issue? It has been around since December 2019, `useQuery` and `useLazyQuery` work as expected so it seems like a deal-breaker for me. I don't want to write try/catch logic around mutations, that's why I use the hooks so I can respond to changes in the UI without side-effect code.

**Comment by webster91** [comment]
It's been almost a year. Why does no one solve this problem?

**Comment by noglitchyo** [comment]
As a workaround, my solution is that since I catch all possible/uncatched errors in the `onError` link and I have no need to forward the errors to other links or whatsoever, I use [```response.errors = undefined```](https://www.apollographql.com/docs/link/links/error/#ignoring-errors) at the end of my error handler.

**Comment by krakzk** [comment]
@noglitchyo Could you please share the onError you have implemented ? I need to get execution terminated at onError in case of a mutation

**Comment by noglitchyo** [comment]
@krakzk Yep, for sure!
```
    const errorLink = onError(({graphQLErrors, networkError, response}) => {
        if (graphQLErrors) {
            graphQLErrors.map(({message, extensions, locations, path}) => {
                //  custom error catching logic
            }
            response.errors = undefined
        }
    }
```
On my end, the "custom error catching logic" do operations depending on extension categories / content of messages, then dispatch an action to my errors reducer. Hope it helps.

**Comment by krakzk** [comment]
@noglitchyo I have a question that is out of the scope of this  open issue, but would you by any chance know a way to show a modal inside the onError by which i can alert the user about the error details
#7200

**Comment by cmnstmntmn** [comment]
hate this lib and react. 
bloated and fully opinionated 👎

**Comment by cmnstmntmn** [comment]
@noglitchyo you have a top catch custom errors, but what about custom actions errors .. you can't keep the logic inside the error link .. c'mon.

**Comment by albertcito** [comment]
I'm try to fix it with [response.errors = undefined](https://www.apollographql.com/docs/link/links/error/#ignoring-errors) but it doesn't work in any case. I have to add the `onError` in each query or mutation :/

**Comment by albertcito** [comment]
It's a big Apollo's fail. Looks like the library has some problem. It has more than 600 open issues and the roadmap for the V3 is from 2018. I don't know if it will be done. About the onError I "fixed" it with a patch:
```typescript
type useMutationFnType<TQuery, TVariables> = (
  baseOptions?: MutationHookOptions<TQuery, TVariables>
) => MutationTuple<TQuery, TVariables>;

function useWithErrorMiddleware<TQuery, TVariables>(
  useMutationFn: useMutationFnType<TQuery, TVariables>,
  baseOptions?: MutationHookOptions<TQuery, TVariables>,
) {
  const options = {
    ...baseOptions,
    onError: () => false,
  };
  return useMutationFn(options);
}
```

So you can use it like. I'm using codegen:

```typescript
const [updatePasword, { loading, error }] = useWithErrorMiddleware(
    useProfileUpdatePasswordMutation,
    {
      errorPolicy: 'all',
      onCompleted: (data) => {
        if (data) {
          const { message, type } = data.profileUpdatePassword.message;
          notification[type]({ message });
        }
      },
    },
  );
```

**Comment by adisrael** [comment]
Found a workaround that worked for me in an old issue [comment](https://github.com/apollographql/apollo-link/issues/855#issuecomment-485764697).

With [apollo-link-error](https://www.apollographql.com/docs/link/links/error/) `return Observable.of(operation);` at the end of the `networkError` logic.

```
import { onError } from "apollo-link-error";

const errorLink = onError(({ graphQLErrors, networkError, operation, forward }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(({ message }) =>
      console.log(`[GraphQL error]: Message: ${message}`)
    );
  if (networkError) {
    console.log(`[Network error]: ${networkError.message}`);
    return Observable.of(operation);
  }
  return forward(operation);
});
```

**Comment by jycouet** [comment]
I think it's going in the right direction @adisrael , but it's not working for me.

I mean, I see the global log `console.log([GraphQL error]...`, but shouldn't it go back also until the mutation? and have result.errors filled?
```
const result = await CreatePeople({
   variables: { fields, accountId: contratId },
});
console.log(`result.errors`, result.errors);
```

Thx for your help

**Comment by cmnstmntmn** [comment]
@jycouet you may have a look over [this issues](https://github.com/apollographql/apollo-client/issues/7565)

if you have `@client fields` .. they might by the reason of not getting the errors.

**Comment by jycouet** [comment]
nope, I don't have any `@client fields`

**Comment by AaronActu** [comment]
Hi team !

Is there some updates ?

I actually have the same problem with @apollo/client ^3.3.11

Here is my snippet : ```js
import { onError } from "@apollo/client/link/error"
import cogoToast from 'cogo-toast'

const handleError = onError(({ graphQLErrors, networkError }) => {
  alert("FUCKING ERRORORORORO")
  if (graphQLErrors)
    graphQLErrors.forEach(({ message }) =>
      cogoToast.error(message)
    );

  if (networkError) cogoToast.warn(networkError.message);
});

const graphqlClient = new ApolloClient({
  uri: "http://localhost:1968/graphql",
  cache: new InMemoryCache(),
  onError: () => handleError
})
```

**Comment by E-Kuerschner** [comment]
Also experiencing this issue (onError not invoked for mutation errors) on @apollo/client v3.3.14

**Comment by valentingavran** [comment]
I'm not sure if I understood this issue correctly, but according to the information I found in the documentation, here's how you should catch the bugs. This works for me without any problems. Maybe this will help someone.

```js
import { ApolloClient } from "apollo-client";
import { setContext } from "apollo-link-context";
import { HttpLink } from "apollo-link-http";
import { ApolloLink } from "apollo-link";
import { onError } from "apollo-link-error";
import { InMemoryCache } from "apollo-cache-inmemory";

const httpLink = new HttpLink({
  uri: "http://localhost:10000"
});

const authLink = setContext((_, { headers }) => {
  // return the headers to the context so HTTP link can read them
  const accessToken = localStorage.getItem("accessToken");
  return {
    headers: {
      ...headers,
      authorization: accessToken
    }
  };
});

const cache = new InMemoryCache();

const apolloClient = new ApolloClient({
  link: ApolloLink.from([
    onError(() => {
      console.log("onError called");
      // Handle errors
     // ...
    }),
    authLink.concat(httpLink)
  ]),
  cache,
});

export default apolloClient;
```

**Comment by Chameleon45** [comment]
Hmmm, weird... Her is my snippet and it works fine only with Observable. Without it I still able to receive errors in console but my client app crashes and outputs an error on page.
```js
const USER_CLIENT = "USER_CLIENT";
const ADMIN_CLIENT = "ADMIN_CLIENT";

const userHttpLink = createHttpLink({
    uri: "https://localhost:5001/user/graphql"
});

const adminHttpLink = createHttpLink({
    uri: "https://localhost:5001/admin/graphql"
});

const errorLink = onError(({ graphQLErrors, networkError, operation }) => {
    if (graphQLErrors) {
        graphQLErrors.map(({ message, locations, path }) => {
            console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`);
            return Observable.of(operation);
        });
        message.error("Oops! Seems something went wrong... Please reload page and try again :)", 5);
    }else if (networkError) {
        console.log(`[Network error]: ${networkError}`);
        return Observable.of(operation);
    }

    return Observable.of(operation);
});

const clientsLink = ApolloLink.split(
    operation => operation.getContext().clientName === USER_CLIENT,
    userHttpLink,
    adminHttpLink
);

const linksChain = from([errorLink, clientsLink]);

const apolloClient = new ApolloClient({
    link: linksChain,
    cache: new InMemoryCache()
});
```

**Comment by Calvin1121** [comment]
I got the same issue, and I found the documentation. But it doesn’t work for me. That’s so weird. 🤷‍♂️

**Comment by wapgear** [comment]
Any updates on this one? I can't believe this issue still not addressed yet

**Comment by oskari** [comment]
Using `return Observable.of(operation)` results on `Missing field 'XXX' while writing result {}` error being printed on console, any idea how to solve this?

**Comment by nhuesmann** [comment]
Hi there. Pinging this thread because I believe I'm running into this issue as well. I'm getting `networkError` when a mutation errors. Could anyone provide a working alternative, or could Apollo advise us on how we should proceed? Thanks.

**Comment by sarink** [comment]
Hello? Apollo, are you there?

**Comment by sarink** [comment]
Hello? Apollo, are you there? Still broken A YEAR LATER in `@apollo/client 3.5.6`...

**Comment by suicaed** [comment]
Still broken.
Used code similar to example from docs:
```
const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(({ message, locations, path }) =>
      console.log(
        `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
      )
    );
  if (networkError) console.log(`[Network error]: ${networkError}`);
});
```

Setting error policy does not work.

The only working way I found is to use 'catch' after client.query():
````
client.query({ query: SomeTestQuery }).catch((error) => console.log(error))
````

But catching errors like this (especially if you're working on a big project) is definitely the way to an asylum.

**Comment by robin-pham** [comment]
Was able to find a solution by overriding the default value of [errorPolicy](https://www.apollographql.com/docs/react/data/error-handling/) specifically for mutations.

Noticed that the line of code that's throwing these errors occurs at https://github.com/apollographql/apollo-client/blob/v3.5.8/src/core/QueryManager.ts#L238. 

So you can simply just set the `errorPolicy` to a non `'none'` value (I opted for `'all'`) at the client level with [defaultOptions](https://www.apollographql.com/docs/react/api/core/ApolloClient/#example-defaultoptions-object). This way you don't have to `try/catch` or use `.mutation().catch()` after each usage since it no longer throws an error, just returns the `error` result.


e.g:
```
import { ApolloClient, InMemoryCache } from '@apollo/client';

const cache = new InMemoryCache();

const client = new ApolloClient({
  cache: cache,
  uri: 'http://localhost:4000/',
  defaultOptions: {
    mutate: {
      errorPolicy: 'all',
    },
  },
});

```

The difference of using `'all'` rather than the default `'none'` wasn't problematic since the codebase I work with never relied on `data` to be `undefined`, and instead relied on the existence of an `error` response.

**Comment by schpet** [comment]
the only workaround that seems to work for me is to provide `onError: () => null` as [described above](https://github.com/apollographql/apollo-client/issues/5708#issuecomment-569754975)

```ts
const [exec] = useMutation(myMutation)
handleThing = () => {
  const result = await exec({onError: () => null, ... })
}
```

confused about how the many apollo-client users work around this? only using `onError` / `onCompleted` callbacks? wrapping all mutation promises with `try { ... } catch (err) { ... }` ?

**Comment by WGTW** [comment]
Still having this issue in projects using @apollo/client 3.4.10 and 3.5.5; this issue was opened over two years ago now, what gives? Even with the posted workarounds it would be worthwhile to fix this bug. Thanks!

**Comment by strblr** [comment]
> Maybe it's time to move away from apollo? 🤔

I eventually moved to urql. No regrets.

**Comment by Cloud11PL** [comment]
Any update regarding this issue?

**Comment by bkiac** [comment]
> defaultOptions: {
>     mutate: {
>       errorPolicy: 'all',
>     },
>   },

I tried this, but this way the `onError` handler is never called

```ts
const [mutate, {error}] = useMutation({onError: (errorFromHandler) => console.log("you can't see me", errorFromHandler)})
console.log("this works", error)
```

**Comment by ryanrhee** [comment]
is this considered a bug or an expected behavior? can someone from the apollo team let us know what we should expect in terms of behavior?

**Comment by brainkim** [comment]
@ryanrhee I am chiming right now let me just read through 40 comments of people talking past each other, one sec.

**Comment by brainkim** [comment]
Hello! Apollo staff here. First of all I want to apologize for the lack of response in this 43 comment, 3 year old issue. I’m sorry that you all had to experience being here. However, I am also glad to see people using this issue as a way to help each other. For instance, I personally appreciate comments likes those of @valentingavran https://github.com/apollographql/apollo-client/issues/5708#issuecomment-824727128, @viiiprock https://github.com/apollographql/apollo-client/issues/5708#issuecomment-569268951 and @Illia-Linevych https://github.com/apollographql/apollo-client/issues/5708#issuecomment-825201861, people who went out of their way to try and help others even after they had solved their own problems. I also appreciate the comment from @kissthom https://github.com/apollographql/apollo-client/issues/5708#issuecomment-588300148 attempting to find the source of the error in what is likely now a very outdated version of the codebase. I can’t respond directly to every comment in this issue, because that would take a lot of time, but I’m glad that this has been a place for people to help each other, even if we are all frustrated when we’re here.

The problem, as best I understand it, is that the `useMutation()` hook, in earlier versions of Apollo Client, threw unhandled errors or promise rejections, and that this was causing problems for developers, especially those who use tools like `react-dev-server`, because the developers who created those tools, thought it would be a good developer experience to throw up an overlay any time it detected an error. Spoiler alert, it’s not.

One thing that complicates this issues is that there are multiple kinds of errors that can be thrown. Network errors, and errors passed to the client from the server, for instance, might bubble up from the link abstraction, which is why people have suggested overriding the flexible link abstraction to ignore the network error.

Other solutions, like adding an `onError()` callback to the `useMutation()` hook, have been added in successive versions, can also squash the error. Additionally, the `useMutation()` `execute()` function returns a promise, which if not caught, could, in various versions, cause an unhandled promise rejection.

In my estimation, I think people in this issue could be talking about any of these kinds of errors. All of these issues have hopefully been remedied in the latest version of Apollo Client, and we encourage you to update when you get the chance.

**In short, I think the problems in this issue have been addressed by the latest version of Apollo Client, and I encourage you to open new issues, and hopefully issues with code snippets, if you think we have not addressed the problems in this issue.**

If you’re just looking for a place to vent about Apollo and GraphQL, I hear you. I feel the same way about of this. Please do not let me closing this issue with comment prevent you from complaining 😊 Thank you for engaging with us!

**Comment by benjamn** [comment]
To be clear, we believe the original `onError` regression should have been addressed by PR #9076, which was first released in v3.5.4. This issue probably should have been closed when that PR was released, so at least we could've found out sooner if it wasn't really fixed. We are (as @brainkim said) happy to continue the discussion here and reopen if necessary!

**Comment by ryanrhee** [comment]
so i'm running on version 3.6.2 and i'm still seeing that the `onError()` link (the one from `import { onError } from "@apollo/client/link/error"`) is not firing when an error happens as a part of a mutation hook.

1. should i open a new issue for this?
2. is this something you thought would be fixed in the latest version of Apollo Client (^3.6) ?

**Comment by ryanrhee** [comment]
@brainkim hi, friendly ping here; happy to open a new issue if this is separate.

**Comment by recurser** [comment]
Another workaround is to provide a custom `fetch` function which transforms the HTTP response code to a `200`:

```typescript
import { ApolloClient } from '@apollo/client'
import { createHttpLink } from '@apollo/client/link/http'

const errorlessFetch = (
  uri: string,
  options: RequestInit,
): Promise<Response> => {
  return new Promise((resolve) => {
    fetch(uri, options).then((response) => {
      // Treat 422s as 200s:
      if (response.status === 422) {
        const { body, ...options } = response
        return resolve(new Response(body, { ...options, status: 200 }))
      }

      return resolve(response)
    })
  })
}

const httpLink = createHttpLink({
  fetch: errorlessFetch,  // <--------------- over-ride fetch()
  uri: 'http://localhost/graphql',
})

export const apiClient = (): ApolloClient<unknown> => {
  return new ApolloClient({
    link: httpLink,
  })
}

```
